### PR TITLE
urlapi: fix PVS warning V560

### PR DESCRIPTION
**Pull request by kvarec**
[CWE-571] V560: A part of conditional expression is always true: qlen.
https://pvs-studio.com/en/docs/warnings/v560/

### CODE REVIEW

File: lib/urlapi.c [hunk 1]
```diff
@@ -1102,10 +1102,10 @@ static CURLUcode parseurl(const char *url, CURLU *u, unsigned int flags)
     hostp = p; /* host name starts here */
 
     /* find the end of the host name + port number */
-    while(*p && !HOSTNAME_END(*p))
+    while(p && *p && !HOSTNAME_END(*p))
```
Comment (bagder) [inline lib/urlapi.c:1105]: How can `p` be NULL here?

Comment (kvarec) [reply lib/urlapi.c:1105]: If schemelen == false, p can be not initialized...

Comment (kvarec) [reply lib/urlapi.c:1105]: Maybe need also 'const char *p = NULL;'?

Comment (bagder) [reply lib/urlapi.c:1105]: Isn't `p` then assigned on line 1100?

Comment (kvarec) [reply lib/urlapi.c:1105]: Oh, see...
But url can be NULL? On release build checking `DEBUGASSERT(url)` is dummy and potential p can be NULL?

Comment (bagder) [reply lib/urlapi.c:1105]: `url` should not be possible to be NULL as input, the assert is there to make sure it dies if that would ever happen anyway.

Comment (kvarec) [reply lib/urlapi.c:1105]:
```c
#if defined(DEBUGBUILD)
#define DEBUGASSERT(x) assert(x)
#else
#define DEBUGASSERT(x) do { } while(0)
#endif
```
assert enabled only for DEBUGBUILD, and on release build if potentially `url == NULL` the application dies, or not?

Comment (bagder) [reply lib/urlapi.c:1105]: Then how does `url` end up NULL  there?

Comment (kvarec) [reply lib/urlapi.c:1105]: Someone will write a code with an bug that will result in `parseurl` call with `url == NULL`.

Comment (bagder) [reply lib/urlapi.c:1105]: The ASSERT is there to catch such possible future mistake (in tests). `parseurl()` is only called from within a single function: `parseurl_and_replace()`.

`parseurl_and_replace()` is called on two places and the URL argument cannot be NULL in either invocation.

If `parseurl()` needs to handle a NULL url then that check should be added explicitly in the beginning of the function. But I don't think it needs to.

Comment (kvarec) [reply lib/urlapi.c:1105]: Ok, i get it.

**Review by bagder — APPROVED** [review]

**Comment by bagder** [comment]
Thanks!

**Pull request closed without merging**